Support Inputs To Splines

danieldresser-ie opened the pull request:
Less fully considered than the corresponding Gaffer PR, but putting it up so you can see this half of it while looking at the Gaffer side.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+		CompoundDataPtr newParametersData = new CompoundData();
+		CompoundDataMap &newParameters = newParametersData->writable();
+
+		for( const auto &i : origParameters )
```
I'm a fan of structured bindings for this sort of thing, so you can have meaningful names instead of `i.first` and `i.second` :

```
for( const auto &[name, value] : origParameters )
```

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+			continue;
+		}
+
+		const auto &origParameters = s.second->parameters();
+		bool hasSplines = false;
+		for( const auto &i : origParameters )
+		{
+			if(
+				runTimeCast<const SplinefColor3fData>( i.second.get() ) ||
+				runTimeCast<const SplineffData>( i.second.get() )
+			)
+			{
+				hasSplines = true;
+				break;
+			}
+		}
+
+		if( !hasSplines )
+		{
+			continue;
+		}
```
What's the purpose of this pre-check? Is it just to avoid the `new CompoundData` when we don't need it? Thinking forwards to your todo on L916, this `continue` seems a bit incompatible with integrating the other conversions. I wonder if the loop over the shaders could call `convertSplines()`, `convertComponents()` etc, passing them each a lambda which will acquire the new parameters on demand? Not sure how important this is now though...

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+			int knotOffset;
+
+			// Insert a conversion shader to handle connection to component
+			auto inserted = currentSplineArrayAdapters.insert( { parameterName, std::make_pair( IECore::InternedString(), 0 ) } );
```
This is another spot where structured bindings can add clarity (over things like `inserted.first->second.second`!).

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
 
+void ShaderNetworkAlgo::convertToOSLConventions( ShaderNetwork *network, int oslVersion )
+{
+	std::map< IECore::InternedString, std::pair< IECore::InternedString, size_t > > currentSplineArrayAdapters;
```
Does this need to be declared all the way up here? It's so far from the point of usage, with no explanation as to what the types mean, so it makes it pretty hard to understand.

danieldresser-ie replied:
I'm not a huge fan of reallocating the map for every shader as a way of documenting its existence, but happy to add a comment where it's defined, or where it's used, or both.  And maybe some of the other changes you've requested will reduce the distance between the beginning of the loop and where it's used.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
```
What if it's not? Feels like we'd want at least a warning?

johnhaddon replied:
Oh, I see we're dealing with that lower down. Feels a bit confusing to separate the two - another thing that could be simplified by the templating suggestion.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+					}
+				}
+				else
+				{
+					knotOffset = getEndPointDuplication( floatSplineData->readable().basis ).first;
+					adapter = new Shader( "Utility/FloatArrayInputAdapter", "osl:shader" );
+
+					const FloatVectorData *splineValuesData = findValues != newParameters.end() ? runTimeCast<const FloatVectorData>( findValues->second.get() ) : nullptr;
+					if( !splineValuesData )
+					{
+						throw IECore::Exception( "Internal failure in convertToOSLConventions - expandSpline did not create values" );
+					}
+
+					const std::vector< float > &splineValues = splineValuesData->readable();
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new FloatData( splineValues[i] );
+						}
+					}
+				}
```
This is just a couple of the branch above, but with different types, right? I think there's enough code here that it's probably worth templating out into a separate function. Long functions have their place, but this one has so much nested control flow that I think it could be a fair bit clearer if it were split into separate parts.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+					);
+				}
+
+				// TODO - does this need to be tagged to be reversible?
+				// ANSWER - it's a special node type, which users shouldn't be creating, so John pointed out
+				// we can probably just remove them all without needing to tag
+				//adapter->blindData()->writable()[ componentConnectionAdapterLabel() ] = g_trueData;
```
If this is the way we're going, then I think we need to change the Gaffer PR to make the adapter more hidden.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+				if( colorSplineData )
+				{
+					knotOffset = getEndPointDuplication( colorSplineData->readable().basis ).first;
+					adapter = new Shader( "Utility/ColorArrayInputAdapter", "osl:shader" );
```
We should add a comment to note that these shaders are actually provided by Gaffer.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+					{ s.first, splineValuesName }
+				) );
+
+				inserted.first->second.first = adapterHandle;
+				inserted.first->second.second = knotOffset;
```
This is where I finally discovered the meaning of the types defined on L718.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -131,6 +132,7 @@ const InternedString g_inParameterName( "in" );
 const InternedString g_outParameterName( "out" );
 const InternedString g_packInParameterNames[4] = { "in1", "in2", "in3", "in4" };
 const boost::regex g_componentRegex( "^(.*)\\.([rgbaxyz])$" );
+const boost::regex g_splineElementRegex( "^(.*)\\[(.*)\\]\\.y(.*)$" );
```
Might be worth tightening this up to require numeric digits in the `[]`?

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+				knotOffset = inserted.first->second.second;
+			}
+
+			// Default is guaranteed not to be used, since from_chars must either set it or set the error code,
+			// but apparently the compiler doesn't know that
+			int elementId = -1;
+			const char *idStart = &destName[splineElementMatch.position(2)];
+			const char *idEnd = idStart + splineElementMatch.length(2);
+			auto elementIdResult = std::from_chars( idStart, idEnd, elementId );
+
+			if( elementIdResult.ec == std::errc::invalid_argument || elementIdResult.ptr != idEnd )
+			{
+				throw IECore::Exception( "Invalid spline point index " + splineElementMatch[2] );
+			}
```
This might just be me, but I really struggle to keep functions this large in my head, and needing 10 lines to do something conceptually simple like convert a string to an int makes it much harder. Personally I doubt performance matters enough here that just `std::stoi( splineElementMatch.str( 2 ) )` wouldn't be perfectly fine. But if we had a tighter regex we could drop all the error checking gumpf here at least.

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+			}
+
+			network->removeConnection( connection );
+			network->addConnection( { connection.source, { adapterHandle, destinationName } } );
```
Do we need to add multiple connections here sometimes, to account for endpoint duplication?

danieldresser-ie replied:
The endPoint duplication that affects the resulting spline shape is done already when we create a SplineData.  The only end point duplication done in IECoreScene::shaderNetworkAlgo is just to account for the extra array positions demanded by OSL for linear and constant.  These values are not actually used - OSL will always treat a 4 component linear array `[ a, b, c, d ]` as a line from `b -> c`, and ignore the values of `a` and `d`.  It is a bit tidier looking to set them to the end values when translating the values, but it doesn't seem worthwhile to create extra connections for the sake of these unused values?

johnhaddon replied:
Ah, cool, that makes sense. Maybe it's a bit misleading to be providing values for the duplicates but not connections though? Maybe just a comment that OSL will ignore the values, so we don't set or connect them?

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+
+	}
+
+	// \todo - it would be a bit more efficient to integrate this, and only traverse the network once,
+	// but I don't think it's worth duplicated the code - fix this up once this call is standard and we
+	// deprecate and remove convertOSLComponentConnections
+	convertOSLComponentConnections( network, oslVersion);
```
Two thoughts :

1. I think an all-in-one function might need to be structured such that it was responsible for the outer loop and then farmed the work out to various `convertFoo( shader, newParameters, newConnections  )` functions. That might allow `convertOSLComponentConnections()` to be implemented similarly, with very little duplication. I think the added structure might also help readability anyway.
2. We could hold this feature till Gaffer 1.3 and Cortex 10.5, giving us carte-blanche to throw out the stuff we don't want.

johnhaddon reviewed (commented):
I think the only comment I made regarding correctness was the query about duplicating connections, but I made a bunch of other comments related to making this more readable, since I have to admit to struggling with it. 

What else is missing before we can consider merging? Tests and USD support?

danieldresser-ie commented:
I was humming and hawing about some of this reorganization to try and better fit what you're looking for on Friday, but I realized it's probably more useful if I at least upload it so you have the option of taking a look if you want to.

danieldresser-ie commented:
I've squashed everything up until when you last glanced at this code, which hopefully makes it easier to review the new stuff, which does involve some rearrangement.

As far as I know, the only outstanding thing is that I haven't tested Appleseed, since we don't have an Appleseed install handy here.

danieldresser-ie commented:
The test failure is probably just due to the Appleseed test not being updated for the last commit where I set the unused extra values for OSL to -1, because it is testing the unused values.  Should be an easy fix.

Also, I forgot to rename concatStrings to concat, but maybe you can figure out what to call intFromStringView and I can rename that at the same time.

danieldresser-ie commented:
OK, I've now updated the string helper names.

I also resolved the Appleseed test, though it did reveal one difference in behaviour:  previously, when the renderer backends were calling expandSplineParameters themselves, they could do that on all shaders, not just OSL shaders.  But when the backend just calls convertToOSLConventions, it only processes OSL shaders.  I think this is reasonable - there are no current examples of where a non-OSL shader would have a spline parameter ... but it's something to think about ( along with it being a good idea for someone with Appleseed handy to briefly test it interactively ).

johnhaddon commented:
Note : I've changed the base of this PR to `RB-10.4` so I don't later get into another mess caused by incorrectly merging to `main`.

danieldresser-ie commented on contrib/IECoreAppleseed/src/IECoreAppleseed/ShaderNetworkAlgo.cpp:
```diff
@@ -65,13 +65,9 @@ namespace ShaderNetworkAlgo
-		shaderNetwork = networkCopy.get();
-	}
+	ShaderNetworkPtr networkCopy = shaderNetwork->copy();
+	IECoreScene::ShaderNetworkAlgo::convertToOSLConventions( networkCopy.get(), 10900 );
```
As discussed, we would have to test not just whether the OSL version used by Appleseed supports component connections, but also whether Appleseed itself supports component connections.  Since our Appleseed support is on a deprecation path anyway, I won't worry about it here ... but it does highlight that driving this functionality based solely on an OSL version input is not totally ideal.

danieldresser-ie commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -789,12 +791,152 @@ void ensureParametersCopy(
+			// doesn't match the input, it is freshly allocated. Once collapseSplineParameters is fully
+			// deprecated, and no longer visible publicly, an internal version of collapseSplineParameters could
+			// just return a non-const new parameter data, or nullptr if no changes are needed.
+			network->setShader( name, std::move( new Shader( shader->getName(), shader->getType(), const_cast< CompoundData *>( collapsed.get() ) ) ) );
```
As discussed, would be a problem if it was doing a remove, but the setShader call for a handle that is already in there should just find the existing handle and repoint it, without invalidating the iterator.

danieldresser-ie commented on include/IECore/StringAlgo.h:
```diff
@@ -180,6 +180,15 @@ bool isUpperCase( const String &s );
+
+/// Concatenate any mixture of strings, string_views, and string literals, efficiently into a new string.
+/// Similar to the std::concat proposed for the standard by P1228, but not yet adopted.
+template<typename ... STRS_T>
```
Renamed STRS_T to StringsFoldType ... not quite sure what the naming convention should be here ... a fold is a fairly unique thing, and I feel like you kind of want a name that calls out that something weird is happening.

danieldresser-ie commented on include/IECore/StringAlgo.inl:
```diff
@@ -171,6 +174,40 @@ inline bool matchInternal( const char *s, const char *&pattern, bool spaceTermin
 	}
 }
 
+template<class STRING_T>
```
Renamed to String ( felt a bit overly generic to me, but matches other functions in the file )

danieldresser-ie commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -789,12 +791,152 @@ void ensureParametersCopy(
+			const IECore::CompoundDataMap &targetParameters = targetShader->parameters();
+
+			int targetSplineKnotOffset = -1;
+			auto findTargetParameter = targetParameters.find( splineName );
```
Renamed to targetParameterIt

danieldresser-ie commented:
Addressed comments in place.  Replied to all of them, except the currentSplineArrayAdapters one which github won't let me reply to, but I did just move the constructor inside the loop for simplicity.

Sorry for the confusion on the issue with constant splines ... it turns out I had gotten used to Gaffer where I always do a full build while testing, and hadn't realized that the last Cortex build I did to test the final duplication commit was just a scons testScene, and it didn't actually install.  So turns out I could reproduce the issue fine ... I think it technically is sort of an OSL bug, but the easy work around was just to revert the last commit, and add a comment about why we don't need to duplicate connections for Y.
